### ISSUE TITLE
DB4 (overhaul db3)

### ISSUE THREAD

**Issue by thenfour** [thenfour/CafeMarcheDB#335]
well this is less of a realistic thing to do, more of a brainstorm of what DB3 should have been or what i can learn.

## Db3 problems 

db3 was aimed at making a MUI datagrid as rich as possible, supporting richer columns than just text/number/date, and adding concepts like tags and foreign single. things like smarter validation, soft deletion, permissions, chip display, quick filtering, and even later search facets all got lumped in, and it ends up a bit of a mess.

to be fair this mess was anticipated because i knew that db columns don't map to application objects, so we would always be dealing with raw db values.

some things in db3 were misguided.

* soft deletion illustrates a fundamental issue. SD works fine, but the problem is db3 tries too hard to include a huge tree of entities. it was a problem so it was elided and then the "enrichment" system was added. it's all hacky. a better approach is to avoid too much inclusion in queries when possible. well soft deletion still may be necessary to add here, but it illustrated the problem.
* putting too much in columns. column (server + client) classes are doing too much. assuming foreignsingle fields work all the same etc, and providing behaviors for literally everything in 1 class (search, render, relate, create, etc). better if these behaviors were plugins.
* payloads. i had this idea of defining a single payload type and using it everywhere. it's my failing as a C++ dev; the better approach is for functions to `<T>` on the payload they require. Like `<T extends Prisma.EventGetPayload<{ select: { name: true } }>>`, which is extremely verbose but it's the typescript way.
* the original plan was an MS access style thing, but it's way too important to make completely custom application stuff, like tweaking the fine details of the UI of events etc.
* the "client intention" stuff is useless. it's a hack to work around default permissions or soft delete behaviors. and almost never used.

so basically, the idea of spicing up the db schema with db3 to provide a rich experience "works", but it's very limited, heavy, and complex for what it does.

## DB4

what db3 really does is try to connect prisma to the application. db4 should no longer stick so close to the db schema. but will be hard to do in a way that i don',t just recreate prisma.

tbh i don't know if it's possible.

**Comment by thenfour** [comment]
More thoughts on the issues with db3 design.

# bundle size is not optimizable

Because of the "do everything related to db" nature of db3, bundle size includes everything db3 may participate in, which is everything. For example MUI datagrid is pretty huge and never used by normal users, but it's included in the bundle because db3 uses it.

I remember being annoyed when realizing i'd have to make separate db3 and db3Client but in fact i should have taken the hint that my design wasn't scalable.

Better to create layers that are modular, testable, and in line with nextjs/react/web architecture.

# needlessly complex

While it does properly encapsulate things like soft delete / visibility permissions / some sanitization, etc, the fact is that it doesn't have much benefit to client code. I learned since then that Prisma payloads should be on demand, not some static thing. Payloads actually do spiral out of control in size very quickly so every component / feature should specify clearly what it needs out of a payload.

Nesting these kinds of payloads deep in a query should be done explicitly by callers, not transparently by some background framework. And because of the needlessly large payloads, it's also often more efficient to make separate async fetches for data than to grab it all in 1 big chunk that mostly won't be seen. For example attendance responses for events are probably best fetched on demand rather than including all responses with every search item in a search result payload.

While that would offload manual work to clients (the thing db3 was invented to eliminate), there are better ways to achieve efficiency / consistency. Provide libraries. Focus on augmenting a schema rather than providing full query caps.

# most gui components don't want to be treated this way

It's rare that the same field will be exposed in exactly the same way in multiple areas of the webiste. So why has it been designed to do exactly that? The concept was clear at the time: a field's behavior shouldn't care where it's being accessed from. Validation of a field can be done near to the conceptual schema. And just one step away from that: if we validate as an email address, then providing a GUI optimized for email input makes sense.

But in practice, these input fields are done in a very custom way per instance, and there was little benefit from providing it deep in the db3 schema.

# Naive schema assumptions

Permissions are set at the column level. Transformations, gui controls, validation, searching, are all done at the db column level. This makes some sense for most fields, but no database schema treats columns like this. The first hint of a problem was in the DateRange of events. Clients don't want to see `startsAt`, `isAllDay`, `durationMilliseconds` etc. They want a date range, and that's what should be validated.

The problem this creates is that clients now need to specially handle any fields that aren't trivial. So it just adds a layer of complexity.

# DB4 should think like an API

A better solution would be to provide utilities in order for callers to do operations efficiently, while not taking away precise control over prisma queries or even raw SQL. We start doing that more and more, using GhostFields instead of structured db3 stuff. And mutations / queries instead of db3 queries & mutations.

The separation between client & server should probably be respected better so the server acts like an API with the client, and specific mutations / queries should be added. Instead of a generic "db3 query" which works for any model, better to just have separate queries as needed.
